src/crypto/CryptoHash.cpp

```diff
@@ -29,27 +29,42 @@ class CryptoHashPrivate
 };
 
 CryptoHash::CryptoHash(CryptoHash::Algorithm algo)
+    : CryptoHash::CryptoHash(algo, false) {}
+
+CryptoHash::CryptoHash(CryptoHash::Algorithm algo, bool hmac)
     : d_ptr(new CryptoHashPrivate())
 {
     Q_D(CryptoHash);
 
     Q_ASSERT(Crypto::initalized());
 
-    int algoGcrypt;
+    int algoGcrypt = -1;
+    unsigned int flagsGcrypt = 0;
 
     switch (algo) {
     case CryptoHash::Sha256:
         algoGcrypt = GCRY_MD_SHA256;
         break;
 
+    case CryptoHash::Sha512:
+        algoGcrypt = GCRY_MD_SHA512;
+        break;
+
     default:
         Q_ASSERT(false);
         break;
     }
 
-    gcry_error_t error = gcry_md_open(&d->ctx, algoGcrypt, 0);
+    if (hmac) {
+        flagsGcrypt |= GCRY_MD_FLAG_HMAC;
+    }
+
+    gcry_error_t error = gcry_md_open(&d->ctx, algoGcrypt, flagsGcrypt);
+    if (error) {
+        qWarning("Gcrypt error (ctor): %s", gcry_strerror(error));
+        qWarning("Gcrypt error (ctor): %s", gcry_strsource(error));
+    }
     Q_ASSERT(error == 0); // TODO: error handling
```

Comment:
@angelsl not sure how we should be handling those errors, but I guess we should at some point. Right now it'll just fail silently unless the application was built in debug mode.

Reply:
Yeah, but I think this PR is starting to be quite large..